design/ceph/admission-controller.md

```diff
@@ -0,0 +1,219 @@
+  - server auth
+```
+Verify with
+` kubectl get csr ${csrName}`
+
+**Sample : Approval of Signed Certificate**
+
+```
+kubectl certificate approve ${csrName}
+```
+
+If it is approved, then we can check the certificate with following command 
+
+```
+$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
+```
+
+* Once approved, a generic secret will be created with the given public and private key which will be later mounted onto the server pod for use.
+
+**Sample : Creating a Secret in Kubernetes**
+```
+kubectl create secret generic ${secret} \
+        --from-file=key.pem=${PRIVATE_KEY_NAME}.pem \
+        --from-file=cert.pem=${PUBLIC_KEY_NAME}.pem
+```
+* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
+* All the above resources will be created in rook-ceph namespace
+
+
+Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.
+
+Below is an excerpt of what a ValidatingWebhookConfig looks like 
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: demo-webhook
+webhooks:
+  - name: webhook-server.webhook-demo.svc
+    clientConfig:
+      service:
+        name: webhook-server
+        namespace: rook-ceph
+        path: "/validate"
+      caBundle: ${CA_PEM_B64}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: [""]
+        apiVersions: ["v1"]
+        resources: ["pods"]
+```
+
+We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.
+
+#### Production
+
+The onus of providing proper certificates and CA bundle for the server would be on the user. The following example would provide how it would work in an OCP cluster environment
+
+##### Openshift Container Platform
+
+![Prod_AC_Image](./rook ac-openshift.png)
+
+As shown in the above diagram, In the OCP cluster, The management of certificates would be completely done by the Openshift Service CA Operator. 
+
+The first step would be to generate the certificates and assign them to the proper components. On annotating any service with 
+
+`service.beta.openshift.io/serving-cert-secret-name:<secret-name>` 
+
+We have to deploy the the annotated service at the start to generate the signed certificates. This can done manually  through the command 
+
+`oc create -f ${SERVICE_NAME}.yaml`
+
+For verifying 
+
+`oc describe ${SERVICE_NAME} -n ${NAMESPACE}`
+
+**Output**
+```
+oc describe service <service-name>
+...
+Annotations:              service.beta.openshift.io/serving-cert-secret-name: <service-name>
+                          service.beta.openshift.io/serving-cert-signed-by: openshift-service-serving-signer@1556850837
+```
+The operator automatically generates a signed serving certificate and key pair into a secret in the same namespace as the service. 
```

Comment:
I thought the operator was just going to look for the cert, rather than generating it? Did you mean the script will do this? Or did you say the operator can only do this for OpenShift and not K8s?

Reply:
Sorry for being unclear, I mean to say that the Openshift Service CA operator is going to be the one which generate and maintain the certificates (https://docs.openshift.com/container-platform/4.1/authentication/certificates/service-serving-certificate.html#understanding-service-serving_service-serving-certificate) . This operator works as a part of OCP and would work only for Openshift not K8s. I will modify the statement here.